### ISSUE TITLE
Raw HTML tags containing ampersand are converted to entities

### ISSUE THREAD

**Issue by certainlyakey** [jgm/pandoc#2410]
Hi! 

If we have an ampersand inside an HTML tag (within markdown input) this tag's symbols `<`, `>` and others get converted to entities and it ends up shown as verbatim code. For example, in a markdown file this (note `&` in src attribute):

	<iframe frameborder="0" style="width: 800px; height: 600px;" src="http://www.mindmeister.com/maps/public_map_shell/466227304?width=800&height=600&z=auto"></iframe>
	<img src="http://i.guim.co.uk/img/static/sys-images/Guardian/Pix/pictures/2014/6/4/1401880969956/How-the-Guardian-reported-009.jpg?w=620&q=85&auto=format&sharp=10&s=37930bdf8f48ce68e8d0848a352b730e" alt="">

gives the following output (Reveal.js):

	<p>&lt;iframe frameborder=&quot;0&quot; style=&quot;width: 800px; height: 600px;&quot; src=&quot;http://www.mindmeister.com/maps/public_map_shell/466227304?width=800&amp;height=600&amp;z=auto&quot;&gt;</iframe></p>
	<p>&lt;img src=&quot;http://i.guim.co.uk/img/static/sys-images/Guardian/Pix/pictures/2014/6/4/1401880969956/How-the-Guardian-reported-009.jpg?w=620&amp;q=85&amp;auto=format&amp;sharp=10&amp;s=37930bdf8f48ce68e8d0848a352b730e&quot; alt=&quot;&quot;&gt;</p>

**Comment by Jmuccigr** [comment]
The ampersands need to be encoded as &amp;. You can type the ampersand into the URL bar in your browser, but a HTML file should have them encoded and shouldn't validate otherwise.

Not sure why pandoc responds as it does, but technically you've not written valid HTML for it.

**Comment by certainlyakey** [comment]
OK, thank you! I thought that an entity in a src attribute would render the URL incorrect and the iframe wouldn't load, but turns out it's other way around. 

The situation usually is like this: you copy the embed code containing the iframe tag from some site. And it turns out thus that it's a problem of a site not converting ampersand to entity in an embed code rather than pandoc-related. For example, Youtube does encodes ampersand, while Mindmeister does not. So thanks for this clarification.

**Comment by simonmichael** [comment]
This caught me by surprise when upgrading to 1.15.1 (from 1.13 I think). 

In markdown files, a simple non-encoded ampersand still works in markdown hyperlinks (`<http://some&where>`), but it now breaks HTML hyperlinks (`<a href="http://some&where">x</a>`), entity-encoding the whole thing.

`<a href="http://some&amp;where">x</a>` works, but is tedious and doesn't look good with Emacs' markdown-mode syntax highlighting. I think it's the norm for tools to accept a non-encoded ampersand in this situation - could this be restored ?

**Comment by simonmichael** [comment]
I tried to find which markdown extension might affect this. Here are some data points using pandoc 1.15.1:
```
source:
<div><a href="http://some&where">x</a></div>

<a href="http://some&where">x</a>

markdown_strict:
<div><a href="http://some&where">x</a></div>
<p>&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a></p>

markdown-markdown_in_html_blocks-native_divs:
<div><a href="http://some&where">x</a></div>
<p>&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a></p>

markdown-markdown_in_html_blocks:
<div>
<p>&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a></div></p>
<p>&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a></p>

markdown-native_divs:
<div>
&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a>
</div>
<p>&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a></p>

markdown:
<div>
&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a>
</div>
<p>&lt;a href=&quot;http://some&amp;where&quot;&gt;x</a></p>
```

Ie a HTML hyperlink wrapped in a div works only with markdown_strict, or markdown with both markdown_in_html_blocks and native_divs disabled. A HTML hyperlink not wrapped in a div doesn't works with any of these markups.

**Comment by jgm** [comment]
`<a href="http://some&where">x</a>` is invalid HTML.
`&` always needs to be encoded.

The reason there's a change is that we changed `htmlTag`
in `Text.Pandoc.Readers.HTML`.  (This is used by the
Markdown reader, too, to recognize raw HTML tags.)

`htmlTag` first uses tagsoup to tokenize according to
the HTML5 algorithm.  This algorithm will allow the
illegal `&`, but give a warning.  What changed in 1.15.1
is that we now have `htmlTag` parser fail if such a
warning is produced.  That is, `htmlTag` is now stricter
about what it will recognize as an HTML tag.

Why the change?  Because in its earlier "lax" mode, `htmlTag`
used to recognize things like `<www.pandoc.org/bla#foo>` as an
HTML tag. The relevant issue is #2277.

The change could be reconsidered, but I do think we need
some way to prevent `<www.pandoc.org/bla#foo>` from being
parsed as an HTML tag!

+++ Simon Michael [Nov 06 15 14:04 ]:

**Comment by simonmichael** [comment]
At the least, it could be mentioned in release notes (I couldn't find it there). I imagine it will impact quite a few people.

**Comment by jgm** [comment]
It's this item in 1.15.0.5 release notes:

    * Markdown/HTML readers:  Avoid parsing partial URLs like
      `<www.pandoc.org/blah#foo>` as HTML tags (#2277).

I didn't realize at the time that it might have wider
effects.

+++ Simon Michael [Nov 07 15 09:54 ]:

**Comment by jmatsushita** [comment]
Hi there,

I just bumped into this problem and it took me a while to figure out what was going on. I also inserted an iframe with a source containing ampersands, and I thought the problem came from the iframe tag first tried to fiddle with extensions, until I tried to create a test document which revealed the real problem being ampersands in ```src``` attributes. It is a tricky one to discover and I can imagine that a number of copy-paster of embed code could bump into it and have trouble troubleshooting it. If it would be possible to display warnings but have a laxer parsing of attribute values in html that would be fantastic.
